src/vulkan-renderer/imgui.cpp

```diff
@@ -114,258 +86,109 @@ ImGUIOverlay::ImGUIOverlay(const wrapper::Device &device, const wrapper::Swapcha
         descriptor_builder.add_combined_image_sampler(m_imgui_texture->sampler(), m_imgui_texture->image_view(), 0)
             .build("ImGUI"));
 
-    // TODO() Use pipeline cache!
-    std::vector<VkAttachmentDescription> attachments(1);
-    attachments[0].format = m_swapchain.image_format();
-    attachments[0].samples = VK_SAMPLE_COUNT_1_BIT;
-    attachments[0].loadOp = VK_ATTACHMENT_LOAD_OP_LOAD;
-    attachments[0].storeOp = VK_ATTACHMENT_STORE_OP_STORE;
-    attachments[0].stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
-    attachments[0].stencilStoreOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
-    attachments[0].initialLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;
-    attachments[0].finalLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;
-
-    VkAttachmentReference color_ref = {};
-    color_ref.attachment = 0;
-    color_ref.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
-
-    VkSubpassDescription subpass_desc = {};
-    subpass_desc.pipelineBindPoint = VK_PIPELINE_BIND_POINT_GRAPHICS;
-    subpass_desc.colorAttachmentCount = 1;
-    subpass_desc.pColorAttachments = &color_ref;
-
-    std::vector<VkSubpassDependency> subpass_deps(1);
-    subpass_deps[0].srcSubpass = VK_SUBPASS_EXTERNAL;
-    subpass_deps[0].srcStageMask = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
-    subpass_deps[0].dstStageMask = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
-    subpass_deps[0].dstAccessMask = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
-
-    spdlog::debug("Creating ImGUI renderpass");
-
-    m_renderpass = std::make_unique<wrapper::RenderPass>(m_device, attachments, subpass_deps, subpass_desc, "ImGUI");
-
+    m_index_buffer = render_graph->add<BufferResource>("imgui index buffer", BufferUsage::INDEX_BUFFER);
+    m_vertex_buffer = render_graph->add<BufferResource>("imgui vertex buffer", BufferUsage::VERTEX_BUFFER);
+    m_vertex_buffer->add_vertex_attribute(VK_FORMAT_R32G32_SFLOAT, offsetof(ImDrawVert, pos));
+    m_vertex_buffer->add_vertex_attribute(VK_FORMAT_R32G32_SFLOAT, offsetof(ImDrawVert, uv));
+    m_vertex_buffer->add_vertex_attribute(VK_FORMAT_R8G8B8A8_UNORM, offsetof(ImDrawVert, col));
+    m_vertex_buffer->set_element_size(sizeof(ImDrawVert));
+
+    m_stage = render_graph->add<GraphicsStage>("imgui stage");
+    m_stage->writes_to(back_buffer);
+    m_stage->reads_from(m_index_buffer);
+    m_stage->reads_from(m_vertex_buffer);
+    m_stage->bind_buffer(m_vertex_buffer, 0);
+    m_stage->uses_shader(*m_vertex_shader);
+    m_stage->uses_shader(*m_fragment_shader);
```

Comment:
I can already see a builder patter applied here. I might do this in an upcoming pull request.

```cpp
m_stage = render_graph->add<GraphicsStage>("imgui stage")
   ->writes_to(back_buffer)
   ->reads_from(m_index_buffer)
   // ...
```

Opinion?

Reply:
I'm indifferent on this - I don't think it would make a whole lot of difference apart from visual.